Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp

```diff
@@ -555,18 +543,21 @@ void IntegratePeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
             FunctionFactory::Instance().createInitialized(fun_str.str());
         boost::shared_ptr<const CompositeFunction> fun =
             boost::dynamic_pointer_cast<const CompositeFunction>(ifun);
-        const Mantid::MantidVec &x = wsProfile2D->readX(i);
-        wsFit2D->dataX(i) = x;
-        wsDiff2D->dataX(i) = x;
-        FunctionDomain1DVector domain(x);
+        const auto &x = wsProfile2D->x(i);
+        wsFit2D->mutableX(i) = x;
+        wsDiff2D->mutableX(i) = x;
+        FunctionDomain1DVector domain(x.rawData());
         FunctionValues yy(domain);
         fun->function(domain, yy);
-        const Mantid::MantidVec &yValues = wsProfile2D->readY(i);
-        for (size_t j = 0; j < numSteps; j++) {
-          wsFit2D->dataY(i)[j] = yy[j];
-          wsDiff2D->dataY(i)[j] = yValues[j] - yy[j];
-        }
 
+        const auto &yValues = wsProfile2D->y(i);
+        auto &fit2DY = wsFit2D->mutableY(i);
+        auto &diff2DY = wsDiff2D->mutableY(i);
+        auto funcValues = yy.toVector();
+        fit2DY = funcValues;
```

Comment:
Why not use `Histogram` arithmetic? You could do something like 
```
wsFit2D->setSharedY(i, wsDiff2D->y(i));
wsFit2D->mutableY(i) -= yy;
```

or something similar.